File: mac/Sources/FocusKit/Services/CSV.swift
import Foundation

/// Just enough CSV for one writer and one reader: quoted fields, doubled quotes.
public enum CSV {
    public static func escape(_ field: String) -> String {
        guard field.contains(where: { $0 == "," || $0 == "\"" || $0 == "\n" }) else { return field }
        return "\"" + field.replacingOccurrences(of: "\"", with: "\"\"") + "\""
    }

    public static func row(_ fields: [String]) -> String {
        fields.map(escape).joined(separator: ",")
    }

    public static func parse(row: String) -> [String] {
        var fields: [String] = []
        var field = ""
        var inQuotes = false
        let characters = Array(row)
        var index = 0

        while index < characters.count {
            let character = characters[index]
            if inQuotes {
                if character == "\"" {
                    // A doubled quote inside a quoted field is a literal quote.
                    if index + 1 < characters.count, characters[index + 1] == "\"" {
                        field.append("\"")
                        index += 1
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(character)
                }
            } else if character == "\"" {
                inQuotes = true
            } else if character == "," {
                fields.append(field)
                field = ""
            } else {
                field.append(character)
            }
            index += 1
        }

        fields.append(field)
        return fields
    }
}

File: windows/Focus.Core/Services/Csv.cs
namespace Focus.Core;

/// Just enough CSV for one writer and one reader: quoted fields, doubled quotes.
public static class Csv
{
    public static string Escape(string field) =>
        field.Any(c => c is ',' or '"' or '\n')
            ? "\"" + field.Replace("\"", "\"\"") + "\""
            : field;

    public static string Row(IEnumerable<string> fields) =>
        string.Join(",", fields.Select(Escape));

    public static List<string> Parse(string row)
    {
        var fields = new List<string>();
        var field = new System.Text.StringBuilder();
        var inQuotes = false;

        for (var i = 0; i < row.Length; i++)
        {
            var c = row[i];
            if (inQuotes)
            {
                if (c == '"')
                {
                    // A doubled quote inside a quoted field is a literal quote.
                    if (i + 1 < row.Length && row[i + 1] == '"') { field.Append('"'); i++; }
                    else inQuotes = false;
                }
                else field.Append(c);
            }
            else if (c == '"') inQuotes = true;
            else if (c == ',') { fields.Add(field.ToString()); field.Clear(); }
            else field.Append(c);
        }

        fields.Add(field.ToString());
        return fields;
    }
}
